### PR TITLE
Add general reference

### PR DESCRIPTION
**Pull request by nstair**
## Description
This PR enables optional passing of a QuantumCircuit object to the Algorithm constructor to act as the initial state preparation circuit.

## User Notes
- [x] Features added
- [ ] Changes to compilation (if any)

## Checklist
- [x] Added/updated tests of new features and included a reference `output.ref` file
- [x] Removed comments in code and input files
- [ ] Documented source code
- [ ] Checked for redundant headers
- [x] Checked for consistency in the formatting of the output file
- [ ] Documented new features in the manual
- [ ] Ready to go!

### CODE REVIEW

**Comment by JonathonMisiewicz** [comment]
I'll start my review once this has approval from Ilias.

File: tests/circ_ref_test.py
Comment (imagoulas) [inline tests/circ_ref_test.py:16]:
```suggestion
        # The FCI energy of the equidistant H4/STO-3G chain with r = 1.5 angs,.
```

**Review by imagoulas — APPROVED** [review]
There is one fragment in abc/algorithm.py that I feel is not right. Please see the corresponding comment and suggestion.

File: src/qforte/ucc/uccnvqe.py
Comment (JonathonMisiewicz) [inline src/qforte/ucc/uccnvqe.py:210]:
```suggestion
```

File: src/qforte/abc/algorithm.py
Comment (JonathonMisiewicz) [inline src/qforte/abc/algorithm.py:64]: Is `trial_state` a standard name for a non-determinantal reference in quantum computing? I find the nomenclature confusing.

**Review by JonathonMisiewicz — COMMENTED** [review]
If I can ask, what's the status of the UCC docs PR? There are some changes I'd like, but perhaps those should wait until then.

File: src/qforte/abc/algorithm.py
Comment (nstair) [reply src/qforte/abc/algorithm.py:64]: This could probably be changed to '_state_prep_type', sometimes we use 'trial state' to describe the general vqe/pqe state in our article, but I will commit the variable name change.

**Comment by nstair** [comment]
> If I can ask, what's the status of the UCC docs PR? There are some changes I'd like, but perhaps those should wait until then.

It is stalled at the moment, mostly because of all the changes we have been making. My plan for order of operations is finish resolving qforte-con issues -> cleanup dead code in all algorithm implementations -> reconstruct docs.

File: src/qforte/abc/algorithm.py
Comment (JonathonMisiewicz) [inline src/qforte/abc/algorithm.py:70]: Do we ever validate that the reference is of the correct length?

**Review by JonathonMisiewicz — COMMENTED** [review]
One last comment. After this is addressed, feel free to merge in.

File: src/qforte/abc/algorithm.py
Comment (nstair) [reply src/qforte/abc/algorithm.py:70]: These lines accomplish this I think.

```
self._nqb = len(self._ref)
self._qb_ham = system.get_hamiltonian()
if self._qb_ham.num_qubits() != self._nqb:
     raise ValueError(f"The reference has {self._nqb} qubits, but the Hamiltonian has {self._qb_ham.num_qubits()}. This is inconsistent.")
```

**Pull request merged**
